[MLIR][test] Run Integration tests for CPU using qemu-aarch64

zhaoshiz opened the pull request:
To run integration tests using qemu-aarch64 on x64 host, below flags are added to the cmake command when building mlir/llvm:

      -DMLIR_INCLUDE_INTEGRATION_TESTS=ON \
      -DMLIR_RUN_ARM_SVE_TESTS=ON \
      -DMLIR_RUN_ARM_SME_TESTS=ON \
      -DARM_EMULATOR_EXECUTABLE="<...>/qemu-aarch64" \
      -DARM_EMULATOR_OPTIONS="-L /usr/aarch64-linux-gnu" \
      -DARM_EMULATOR_MLIR_CPU_RUNNER_EXECUTABLE="<llvm_arm64_build_top>/bin/mlir-cpu-runner-arm64" \
      -DARM_EMULATOR_LLI_EXECUTABLE="<llvm_arm64_build_top>/bin/lli" \
      -DARM_EMULATOR_UTILS_LIB_DIR="<llvm_arm64_build_top>/lib"

The last three above are prebuilt on, or cross-built for, an aarch64 host.

Some tests marked with 'UNSUPPORTED: target=aarch64{{.*}}' are still run when configured with ARM_EMULATOR_EXECUTABLE and the default target is not aarch64.
A lit config feature 'mlir_arm_emulator' is added in mlir/test/lit.site.cfg.py.in and to UNSUPPORTED list of such tests.

zhaoshiz committed:
[MLIR][Vector] Add a test of <2x[4]xf32> type on lowering multireduction

zhaoshiz committed:
Rename functions in test for better clarity

zhaoshiz committed:
Merge branch 'llvm:main' into mlir-tests

zhaoshiz committed:
Merge branch 'llvm:main' into mlir-tests

zhaoshiz committed:
[MLIR][test] Update CPU Integration Tests to Run Them Using QEMU

To run ArmSVE ArmSME integration tests using qemu-aarch64 on an x86_64 host,
add below flags to the cmake command when building mlir/llvm:

  -DMLIR_INCLUDE_INTEGRATION_TESTS=ON \
  -DMLIR_RUN_ARM_SVE_TESTS=ON \
  -DMLIR_RUN_ARM_SME_TESTS=ON \
  -DARM_EMULATOR_EXECUTABLE="<...>/qemu-aarch64" \
  -DARM_EMULATOR_OPTIONS="-L /usr/aarch64-linux-gnu" \
  -DARM_EMULATOR_MLIR_CPU_RUNNER_EXECUTABLE="<llvm_arm64_build_top>/bin/mlir-cpu-runner-arm64" \
  -DARM_EMULATOR_LLI_EXECUTABLE="<llvm_arm64_build_top>/bin/lli" \
  -DARM_EMULATOR_UTILS_LIB_DIR="<llvm_arm64_build_top>/lib"

The last three above are prebuilt on, or cross-built for, an aarch64 host.

This patch adds 'native_' prefix to mlir utils shared lib names in tests at
loctions: mlir/test/Integration/Dialect/[Linalg|Vector]/CPU/[ArmSME|ArmSVE],
e.g.: %mlir_runner_utils --> %native_mlir_runner_utils. Substitution rules
are added in lit.local.cfg files.

zhaoshiz committed:
[MLIR][test] Update SparseTensor CPU Integration tests to run them on qemu-aarch64

Tests are changed using sed commands below:

sed -i '/\/\/\ DEFINE:\ %{run_libs}/a \/\/\ DEFINE:\ %{run_libs_sve}\ =\ -shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext' Integration/Dialect/SparseTensor/CPU/*.mlir
sed -i 's/"+sve"\ %{run_opts}\ %{run_libs}/"+sve"\ %{run_opts}\ %{run_libs_sve}/' Integration/Dialect/SparseTensor/CPU/*.mlir

XFAILED {run_sve} part of sparse_sum_bf16.mlir and sparse_tanh.mlir as
'UNSUPPORTED: target=aarch64{{.*}}' is not working properly.

MacDue commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/lit.local.cfg:
```diff
@@ -7,3 +7,60 @@ if not config.mlir_run_arm_sme_tests:
 if sys.platform == "win32":
     config.unsupported = True
+
+mlir_runner_utils_file_name = "/libmlir_runner_utils" + config.llvm_shlib_ext
+mlir_c_runner_utils_file_name = "/libmlir_c_runner_utils" + config.llvm_shlib_ext
+mlir_arm_runner_utils_file_name = "/libmlir_arm_runner_utils" + config.llvm_shlib_ext
+arm_sme_abi_shlib_file_name = "/libmlir_arm_sme_abi_stubs" + config.llvm_shlib_ext
+if config.arm_emulator_utils_lib_dir:
+    # Use mlir runtime libs from the location passed by -DARM_EMULATOR_UTILS_LIB_DIR.
+    config.substitutions.append(
+        (
+            "%native_mlir_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_mlir_c_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_c_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (   "%native_mlir_arm_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_arm_runner_utils_file_name
+        )
+    )
```
Can this be moved to a common lit config? I see that these are duplicated for the SVE config too.

zhaoshiz replied:
moved to test/Integration/lit.local.cfg

MacDue commented on mlir/test/Integration/Dialect/Linalg/CPU/ArmSVE/lit.local.cfg:
```diff
@@ -7,3 +8,45 @@ if not config.mlir_run_arm_sve_tests:
 if sys.platform == "win32":
     config.unsupported = True
+
+mlir_runner_utils_file_name = "/libmlir_runner_utils" + config.llvm_shlib_ext
+mlir_c_runner_utils_file_name = "/libmlir_c_runner_utils" + config.llvm_shlib_ext
+mlir_arm_runner_utils_file_name = "/libmlir_arm_runner_utils" + config.llvm_shlib_ext
+if config.arm_emulator_utils_lib_dir:
+    # Use mlir runtime libs from the location passed by -DARM_EMULATOR_UTILS_LIB_DIR.
+    config.substitutions.append(
+        (
+            "%native_mlir_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_mlir_c_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_c_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (   "%native_mlir_arm_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_arm_runner_utils_file_name
+        )
+    )
```
Duplicate configuration.

zhaoshiz replied:
moved to test/Integration/lit.local.cfg

MacDue commented on mlir/test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg:
```diff
@@ -18,5 +35,12 @@ if config.riscv_vector_emulator_executable:
+    emulation_cmd = config.arm_emulator_executable
+    if config.arm_emulator_options:
+        emulation_cmd = emulation_cmd + " " + config.arm_emulator_options
+    emulation_cmd = emulation_cmd + " " + lli_cmd + " --march=aarch64 -mattr=+neon "
```
Why just `-mattr=+neon` here?

c-rhodes replied:
also not clear to me what the rationale here is. I don't believe `arm_emulator_executable` was intended to mean run all the integration tests under emulation, it was intended to support running the integration tests for ISAs like SVE and SME enabled under
- MLIR_RUN_ARM_SVE_TESTS
- MLIR_RUN_ARM_SME_TESTS

where hardware isn't yet widely available. I think we need to mindful here because this will mean these tests are unnecessarily run under emulation on AArch64 hosts and will take longer to run.

zhaoshiz replied:
Reverted changes to test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg and sinked substitutions rule of `%mlir_native_utils_lib_dir` to test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg.

aartbik commented:
LG for sparse changes

zhaoshiz committed:
Address review comments

Moved duplicated substitution rules to test/Integration/lit.local.cfg.

Reverted changes to test/Integration/Dialect/LLVMIR/CPU/lit.local.cfg and sinked
substitutions rule of %mlir_native_utils_lib_dir to
test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg.

Added feature 'mlir_arm_emulator' to UNSUPPORT sparse_sum_bf16.mlir and
sparse_tanh.mlir in test/Integration/Dialect/SparseTensor/CPU.

zhaoshiz commented on mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:
```diff
@@ -27,6 +28,7 @@
 // RUN: %{compile} | %{run} | FileCheck %s
 //
 // Do the same run, but now with  VLA vectorization.
+// XFAIL: {{.*}}
```
changed to UNSUPPORTED: mlir_arm_emulator

zhaoshiz commented on mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_tanh.mlir:
```diff
@@ -28,6 +29,7 @@
 // RUN: %{compile} | %{run} | FileCheck %s
 //
 // Do the same run, but now with  VLA vectorization.
+// XFAIL: {{.*}}
```
changed to UNSUPPORTED: mlir_arm_emulator

MacDue commented on mlir/test/Integration/Dialect/SparseTensor/CPU/block.mlir:
```diff
@@ -10,9 +10,10 @@
 // DEFINE: %{compile} = mlir-opt %s --sparsifier="%{sparsifier_opts}"
 // DEFINE: %{compile_sve} = mlir-opt %s --sparsifier="%{sparsifier_opts_sve}"
 // DEFINE: %{run_libs} = -shared-libs=%mlir_c_runner_utils,%mlir_runner_utils
+// DEFINE: %{run_libs_sve} = -shared-libs=%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext,%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext
```
Why can't this use the substitutions you've defined? Same question for the other parse tests.
```suggestion
// DEFINE: %{run_libs_sve} = -shared-libs=%native_mlir_runner_utilst,%native_mlir_c_runner_utils
```

zhaoshiz replied:
ah, totally missed this part

MacDue commented on mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:
```diff
@@ -29,7 +30,7 @@
 // RUN: %if mlir_arm_sve_tests %{ %{compile_sve} | %{run_sve} | FileCheck %s %}
 
-// UNSUPPORTED: target=aarch64{{.*}}
+// UNSUPPORTED: mlir_arm_emulator
```
I see now that these tests generally don't compile for AArch64 (unrelated to emulation). So, I think this should be something like `UNSUPPORTED: mlir_arm_emulator, target=aarch64{{.*}}`. 

i.e. supported under emulation and real hardware.

zhaoshiz replied:
ok, I added back target=aarch64{{.*}}

MacDue reviewed (commented):
I've tried this locally and for my use cases everything still seems to work :slightly_smiling_face:, so if this allows people to work on x86 too I'm happy with the changes. Just a few more comments/questions:

zhaoshiz committed:
Update SparseTensor CPU tests

Add back 'target=aarch64{{.*}}' to UNSUPPORTED list of sparse_sum_bf16.mlir and
sparse_tanh.mlir.

Replace lengthy '%mlir_native_utils_lib_dir/libmlir_runner_utils%shlibext' and
'%mlir_native_utils_lib_dir/libmlir_c_runner_utils%shlibext' with:
'%native_mlir_runner_utils' and '%native_mlir_c_runner_utils'.

banach-space commented on mlir/test/Integration/lit.local.cfg:
```diff
@@ -58,3 +51,61 @@ aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
+        )
+    )
+    config.substitutions.append(
+        (   "%native_mlir_arm_runner_utils",
+            config.arm_emulator_utils_lib_dir + mlir_arm_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_arm_sme_abi_shlib",
+            # Use passed Arm SME ABI routines, if not present default to stubs.
+            config.arm_sme_abi_routines_shlib or (config.arm_emulator_utils_lib_dir + arm_sme_abi_shlib_file_name)
+        )
+    )
+else:
+    config.substitutions.append(
+        (
+            "%native_mlir_runner_utils",
+            config.mlir_lib_dir + mlir_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_mlir_c_runner_utils",
+            config.mlir_lib_dir + mlir_c_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_mlir_arm_runner_utils",
+            config.mlir_lib_dir + mlir_arm_runner_utils_file_name
+        )
+    )
+    config.substitutions.append(
+        (
+            "%native_arm_sme_abi_shlib",
+            # Use passed Arm SME ABI routines, if not present default to stubs.
+            config.arm_sme_abi_routines_shlib or (config.mlir_lib_dir + arm_sme_abi_shlib_file_name)
+        )
+    )
```
Perhaps I'm missing something, but couldn't this block be rewritten as:
```python
if config.arm_emulator_utils_lib_dir:
  base_dir = config.arm_emulator_utils_lib_dir
else:
  base_dir = config.mlir_lib_dir
  
config.substitutions.append(("%native_mlir_runner_utils", base_dir + mlir_runner_utils_file_name))
config.substitutions.append((...)))
```
? This way the `if`/`else` blocks would get nice and small.

banach-space commented on mlir/test/Integration/lit.local.cfg:
```diff
@@ -58,3 +51,61 @@ aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
 # * ARM_EMULATOR_UTILS_LIB_DIR - AArch64 native utilites library to support
 #                                cross-compilation.
```
The comment on L44-L52 can be moved above this line:
```python
aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
```

Basically, everything that follows is in preparation for possible emulation. I would also move the block that starts with `if config.arm_emulator_utils_lib_dir:` into a function to match `configure_aarch64_mcr_cmd`.

This will most likely require some updates to the comments, but will lead to a much cleaner config.

zhaoshiz replied:
Thanks for the suggestion, I've refactored the substitutions.

banach-space commented on mlir/test/Integration/Dialect/SparseTensor/CPU/sparse_sum_bf16.mlir:
```diff
@@ -29,7 +30,7 @@
 // RUN: %if mlir_arm_sve_tests %{ %{compile_sve} | %{run_sve} | FileCheck %s %}
 
-// UNSUPPORTED: target=aarch64{{.*}}
+// UNSUPPORTED: target=aarch64{{.*}}, mlir_arm_emulator
```
Why is `mlir_arm_emulator`? Could you add a note in the summary? Thanks!

zhaoshiz replied:
I added comments in mlir/test/lit.site.cfg.py.in and the summary.

zhaoshiz committed:
Update for code clarity

Refactored mlir/test/Integration/lit.local.cfg and added comments about new
config feature 'mlir_arm_emulator' in mlir/test/lit.site.cfg.py.in.

zhaoshiz committed:
Merge branch 'llvm:main' into mlir-tests

banach-space commented on mlir/test/Integration/lit.local.cfg:
```diff
@@ -57,4 +64,9 @@ aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
 #                                cross-compilation.
+
+# Check for above options and set correct executables and runtime libs
+aarch64_mcr_cmd = configure_aarch64_mcr_cmd()
```
I would take this one step further and refactor `configure_aarch64_mcr_cmd` to also run:
```python
config.substitutions.append(("%mcr_aarch64_cmd", aarch64_mcr_cmd))
```

This way both methods would be more consistent and the overall code hygiene would improve :)

banach-space reviewed (approved):
This is a very nice improvement, thank you! LGTM

I've left one more small suggestion, but that's just a nit, so feel free to ignore. Please wait for @MacDue to approve before landing.

Thanks again for contributing! 🙏🏻

zhaoshiz committed:
Move substitution of "%mcr_aarch64_cmd" to configure_aarch64_mcr_cmd(), NFC

zhaoshiz committed:
Merge branch 'llvm:main' into mlir-tests

c-rhodes commented on mlir/test/Integration/Dialect/SparseTensor/CPU/lit.local.cfg:
```diff
@@ -11,6 +11,12 @@ if config.mlir_run_arm_sve_tests:
     config.substitutions.append(
         ("%VLA_ARCH_ATTR_OPTIONS", '--march=aarch64 --mattr="+sve"')
     )
+    config.substitutions.append(
+        (
+            "%mlir_native_utils_lib_dir",
+            config.arm_emulator_utils_lib_dir or config.mlir_lib_dir,
+        )
+    )
```
this is no longer used? (and below at the end of the `else`)

zhaoshiz replied:
nice catch. I also updated comments about UNSUPPORTing the tests on Windows.

c-rhodes reviewed (approved):
left one minor but otherwise LGTM, thanks for fixing this it's a nice improvement!

MacDue reviewed (approved):
LGTM, this is nice! Please address @c-rhodes's comment first though :smile:

zhaoshiz committed:
Remove unused substitution "%mlir_native_utils_lib_dir" in SparseTensor/CPU/lit.local.cfg

Also update comments about UNSUPPORTing the test on Windows.

zhaoshiz committed:
Merge branch 'llvm:main' into mlir-tests
